parsl/executors/high_throughput/executor.py

```diff
@@ -730,6 +730,26 @@ class BlockInfo:
 
         managers = self.connected_managers()
         block_info: Dict[str, BlockInfo] = {}
+
+        # First populate the list of blocks with blocks known to the provider
+        # mechanism. Then update that list with information from the
+        # interchange which will include idle times and task counts. Blocks
+        # with no interchange registration (for example because they are still
+        # in a batch queue) will be regarded as idle for a long time (so will
+        # be preferred when killing most idle blocks) in preference to any
+        # blocks which have already registered.
+
+        provider_blocks = self.status()
+        for (b_id, job_status) in provider_blocks.items():
+
+            # skip blocks that have already finished...
+            if job_status.state in TERMINAL_STATES:
+                continue
+
+            if b_id not in block_info:
+                block_info[b_id] = BlockInfo(tasks=0, idle=float('inf'))
```

Comment:
`provider_blocks` is a dictionary, yeah?  If so, then `b_id` is guaranteed to be unique through the loop.  Coupled with the fact that `block_info` is empty when the loop starts, the if-conditional is a waste.  Could merge it all into the instantiation:
```python
block_info = {
  block_id: BlockInfo(tasks=0, idle=float('inf')
  for block_id, job_status in provider_blocks.items()
  if job_status.state not in TERMINAL_STATES
}
```
Later on, `block_info` keys are set to zero on first use as well, so could make that structure a `defaultdict`:
```python
block_info: Dict[str, BlockInfo = defaultdict(lambda: BlockInfo(0, float('inf')))
for b_id, job_status in provider_blocks.items():
    # skip blocks that have already finished...
    if job_status.state in TERMINAL_STATES:
        continue

    # the defaultdict logic is now the authority on "empty BlockInfo" objects
    block_info[b_id]
```
And could remove the if-branch in the following for-loop as well.